eng/targets/VisualStudio.targets

```diff
@@ -47,57 +50,56 @@
         <Path>%(ReferenceCopyLocalPaths.Identity)</Path>
-      </_AssetsByPackageId>
-
-      <_AssetsByPackageId Remove="@(_ExcludedPackageId)" />
+      </_RuntimeAssetsByPackageId>
     </ItemGroup>
 
-    <!-- Flow metadata set on NuGetPackageToIncludeInVsix to VSIXCopyLocalReferenceSourceItem -->
-    <JoinItems Left="@(_AssetsByPackageId)" LeftKey="" LeftMetadata="*"
+    <JoinItems Left="@(_RuntimeAssetsByPackageId)" LeftKey="" LeftMetadata="*"
                Right="@(NuGetPackageToIncludeInVsix)" RightKey="" RightMetadata="*">
-      <Output TaskParameter="JoinResult" ItemName="_AssetsWithMetadata" />
+      <Output TaskParameter="JoinResult" ItemName="_RuntimeAssetsWithMetadata" />
     </JoinItems>
 
     <ItemGroup>
       <!-- Include the assets in the VSIX -->
-      <VSIXCopyLocalReferenceSourceItem Include="@(_AssetsWithMetadata->'%(Path)')">
+      <VSIXCopyLocalReferenceSourceItem Include="@(_RuntimeAssetsWithMetadata->'%(Path)')">
         <ForceIncludeInVsix>true</ForceIncludeInVsix>
         <Private>true</Private>
-        <Ngen Condition="'%(_AssetsWithMetadata.Optimization)' == 'true'">true</Ngen>
-        <NgenArchitecture Condition="'%(_AssetsWithMetadata.Optimization)' == 'true'">All</NgenArchitecture>
-        <NgenPriority Condition="'%(_AssetsWithMetadata.Optimization)' == 'true'">3</NgenPriority>
+        <Ngen Condition="'%(_RuntimeAssetsWithMetadata.Optimization)' == 'true'">true</Ngen>
+        <NgenArchitecture Condition="'%(_RuntimeAssetsWithMetadata.Optimization)' == 'true'">All</NgenArchitecture>
+        <NgenPriority Condition="'%(_RuntimeAssetsWithMetadata.Optimization)' == 'true'">3</NgenPriority>
       </VSIXCopyLocalReferenceSourceItem>
 
       <!-- Add PkgDef* items for assets that specify PkgDefEntry -->
-      <PkgDefBindingRedirect Include="@(_AssetsWithMetadata->'%(Path)')" Condition="'%(_AssetsWithMetadata.PkgDefEntry)' == 'BindingRedirect' and %(_AssetsWithMetadata.AssetType) == 'runtime'" />
-      <PkgDefCodeBase Include="@(_AssetsWithMetadata->'%(Path)')" Condition="'%(_AssetsWithMetadata.PkgDefEntry)' == 'CodeBase' and %(_AssetsWithMetadata.AssetType) == 'runtime'" />
+      <PkgDefBindingRedirect Include="@(_RuntimeAssetsWithMetadata->'%(Path)')" Condition="'%(_RuntimeAssetsWithMetadata.PkgDefEntry)' == 'BindingRedirect'" />
+      <PkgDefCodeBase Include="@(_RuntimeAssetsWithMetadata->'%(Path)')" Condition="'%(_RuntimeAssetsWithMetadata.PkgDefEntry)' == 'CodeBase'" />
+
+      <!-- Check that NuGetPackageToIncludeInVsix doesn't list unreferenced package -->
+      <_UnknownAssets Include="@(NuGetPackageToIncludeInVsix)" />
+      <_UnknownAssets Remove="@(_RuntimeAssetsWithMetadata)" />
     </ItemGroup>
+
+    <Error Text="Packages specified by NuGetPackageToIncludeInVsix are not referenced by the project or have no runtime assets applicable to $(TargetFramework): @(_UnknownAssets)"
+           Condition="'@(_UnknownAssets)' != ''"/>
```

Comment:
Something to treat for a follow up PR, but I wonder if it's worth having the "include in VSIX" just be a flag on the <PackageReference>. I did a quick peek through history and this being a separate item might have been right around when we first moved to SDK-style projects or earlier, so it's entirely possible we had done that either as a hack or at the time we still had project.json....

Reply:
I thought about that. Didn't want to delay the fix, but worth following up.